### PR TITLE
fix(Executor): Fix segfault if callback group is deleted during rmw_wait

### PR DESCRIPTION
**Pull request by jmachowinski**
Fix for https://github.com/ros2/rclcpp/issues/2664 in rolling

Note, this is not my preferred fix for this. But I would like to merge this now,
to get a fix into jazzy ASAP.

@alsora @mjcarroll fyi

### CODE REVIEW

**Review by mjcarroll — APPROVED** [review]

**Comment by mjcarroll** [comment]
Not ideal, but does the job.  Thanks.

File: rclcpp/src/rclcpp/executor.cpp [hunk 1]
```diff
@@ -729,18 +729,37 @@ Executor::wait_for_work(std::chrono::nanoseconds timeout)
...
       this->collect_entities();
     }
+
+    auto callback_groups = this->collector_.get_all_callback_groups();
+    cbgs.resize(callback_groups.size());
+    for(const auto &w_ptr : callback_groups)
+    {
+      auto shr_ptr = w_ptr.lock();
+      if(shr_ptr)
+      {
```
Comment (christophebedard) [inline rclcpp/src/rclcpp/executor.cpp:750]:
```suggestion
    for (const auto & w_ptr : callback_groups) {
      auto shr_ptr = w_ptr.lock();
      if (shr_ptr) {
```

**Comment by wjwwood** [comment]
We talked about a potentially better fix for this in rolling (versus Jazzy) at the client library working group, but I would be ok with a temporary fix in rolling as well.

I also mentioned on the jazzy fix that it would be nice to have a regression test of some sort, but that's more important for rolling than jazzy imo.

File: rclcpp/src/rclcpp/executor.cpp [hunk 2]
```diff
@@ -729,18 +729,35 @@ Executor::wait_for_work(std::chrono::nanoseconds timeout)
   // Clear any previous wait result
   this->wait_result_.reset();
 
+  // we need to make sure that callback groups don't get out of scope
+  // during the wait. As in jazzy, they are not covered by the DynamicStorage,
+  // we explicitly hold them here as a bugfix
```
Comment (fujitatomoya) [inline rclcpp/src/rclcpp/executor.cpp:734]: i would add `TODO` if we already have a plan to proper fix in the future not to miss, this does not block the PR.

**Review by fujitatomoya — APPROVED** [review]
lgtm with green CI, unit test would be ideal.

**Review by alsora — APPROVED** [review]

**Comment by ahcorde** [comment]
Pulls: ros2/rclcpp#2683
Gist: https://gist.githubusercontent.com/ahcorde/fbbd8e165d63b1398346b46e40e02a57/raw/1e348604284103bcceb6dc009bd30720eb3e7ecc/ros2.repos
BUILD args:   --packages-above-and-dependencies rclcpp --packages-above-and-dependencies rclcpp
TEST args:  --packages-above rclcpp --packages-above rclcpp
ROS Distro: rolling
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/14874
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/22186/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/16438/)
* Linux-rhel [Build Status](http://ci.ros2.org/job/ci_linux-rhel/1707/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/22936/)

**Comment by jmachowinski** [comment]
@ahcorde  @alsora  can we merge ?

**Comment by fujitatomoya** [comment]
https://ci.ros2.org/job/ci_windows/22936/msbuild/new/ are unrelated.

**Pull request merged**